naga/src/front/wgsl/parse/mod.rs

```diff
@@ -2433,4 +2454,43 @@ impl Parser {
         }
         Ok(brace_nesting_level + 1)
     }
+
+    fn diagnostic_control<'a>(&self, lexer: &mut Lexer<'a>) -> Result<(), Error<'a>> {
+        // TODO(maybe): self.push_rule_span(Rule::???, lexer)
+        // TODO: I see `*_argument` helpers, but they're not used here. Guessing it's
+        // more valuable to make something easy to validate against the WGSL spec.?
+        lexer.expect(Token::Paren('('))?;
+        let (severity_control_name, severity_control_name_span) = lexer.next_ident_with_span()?;
+        match severity_control_name {
+            "error" | "warning" | "info" | "off" => (/* neat */),
+            _ => todo!(
+                "diagnostic for invalid severity control name {:?}",
+                severity_control_name
+            ),
+        }
+        lexer.expect(Token::Separator(','))?;
+        let (diagnostic_rule_name, diagnostic_rule_name_span) = lexer.capture_span(|lexer| {
+            // TODO: ensure that errors are good
+            let mut segments = Vec::new();
```

Comment:
Maybe use `Vec::with_capacity(2)`?

Reply:
I was actually thinking of replacing this with an `ArrayVec`, pending tests with stack usage. But yes, failing that, setting capacity would be good.

Reply:
Invalidated by my latest force-push, which does the extract of a single diagnostic name token structurally instead.